news/_posts/2019-12-25-prestashop-in-2019-and-beyond-part-3-the-future-architecture.md

```diff
@@ -0,0 +1,269 @@
+- **Features:** How powerful it is, how many things it can do
+
+An ideal software should be good in all four. But let's be honest: in the real world, it's not possible to be excel in everything at once. In order really perform at a high level, some trade-offs need to be made, and it usually means focusing on a smaller number of things.
+
+Much the same as the ["Fast-Good-Cheap, pick two" problem](https://fastgood.cheap/), in my opinion, software can only be realistically optimized for three of these quality axes at a time, and one must always be sacrificed in the interest of the others. Here's a graphical representation:
+
+![Choose your trade-off][architecture-tradeoffs]
+
+If we are forced to choose between one of these archetypes, which one would we be rather be? Not buggy, that's for sure. Not proprietary either, because PrestaShop is an open source platform. We have to choose between having very few features but do them really well, or have many features at the cost of a steeper learning curve.
+
+## Our proposal for a future Architecture
+
+Based on the everything that we stated before, here's what we came up with:
+
+![Schema of the Future Architecture][future-architecture-overview]
+{: .text-center }
+
+The future architecture is based on 5 key elements:
+
+1. Core Domain
+2. Front-end applications
+3. Contracts and Tools (SDKs)
+4. APIs
+5. Extensions
+
+The **Core** sits in the back-end and is at the center of it all. It accounts for all the business needs and use cases that PrestaShop is capable of doing (managing products, shopping carts, orders, etc.). The Core is _domain-oriented_, meaning that it is built around business use cases, expressed in an [ubiquitous language](https://enterprisecraftsmanship.com/posts/ubiquitous-language-naming/). The Core is also master of its own Domain; in  order to let it be the guardian of system-wide coherence, it has to be isolated from other services and be the only one capable of performing state transitions. Other services can only interact with the Core through well-defined interfaces, that we call the Core API. Incidentally, this also means that Core behavior can be _extended_, not _modified_, by other services–or at least not in a way that it puts system coherence at risk.
+
+In the future architecture, the Back Office (BO) and the Front Office (FO) are independent **front-end applications**, each one running entirely on the browser. They are fully component-based (our framework of choice is VueJs), and built using separate toolsets called **Software Development Kits** (SDKs): one for the FO, one for the BO. These SDKs would not only include reusable components, but also bidirectional communication channels based on stable **contracts**, both within the front-end application (events) as well as with the back-end. The BO SDK also integrates the UI Kit, which provides an uniform style for the whole Back Office.
```

Comment:
Shall we provide an external source for VueJS ?

Reply:
```suggestion
In the future architecture, the Back Office (BO) and the Front Office (FO) are independent **front-end applications**, each one running entirely on the browser. They are fully component-based (our framework of choice is [VueJs](https://vuejs.org/)), and built using separate toolsets called **Software Development Kits** (SDKs): one for the FO, one for the BO. These SDKs would not only include reusable components, but also bidirectional communication channels based on stable **contracts**, both within the front-end application (events) as well as with the back-end. The BO SDK also integrates the UI Kit, which provides an uniform style for the whole Back Office.
```